File: fraiseql-php/examples/EcommerceSchema.php
<?php

declare(strict_types=1);

namespace FraiseQL\Examples;

use FraiseQL\Attributes\GraphQLType;
use FraiseQL\Attributes\GraphQLField;
use FraiseQL\StaticAPI;
use FraiseQL\SchemaFormatter;

/**
 * E-commerce schema example demonstrating complex nested types.
 *
 * This example shows how to:
 * - Define complex nested relationships
 * - Use TypeBuilder for programmatic schema construction
 * - Combine attribute-based and builder-based definitions
 * - Handle lists and nullable fields
 */

// Define domain types using attributes
#[GraphQLType(name: 'Category', description: 'Product category')]
final class Category
{
    #[GraphQLField(type: 'Int', description: 'Category ID')]
    public int $id;

    #[GraphQLField(type: 'String', description: 'Category name')]
    public string $name;

    #[GraphQLField(type: 'String', nullable: true)]
    public ?string $description;
}

#[GraphQLType(name: 'Product', description: 'Physical product')]
final class Product
{
    #[GraphQLField(type: 'Int', description: 'Product ID')]
    public int $id;

    #[GraphQLField(type: 'String', description: 'Product name')]
    public string $name;

    #[GraphQLField(type: 'String', description: 'Product description')]
    public string $description;

    #[GraphQLField(type: 'Float', description: 'Price in USD')]
    public float $price;

    #[GraphQLField(type: 'Int', description: 'Stock quantity')]
    public int $stock;

    #[GraphQLField(type: 'Category', description: 'Product category')]
    public Category $category;

    #[GraphQLField(type: 'Boolean', description: 'Whether product is active')]
    public bool $active;
}

#[GraphQLType(name: 'OrderItem', description: 'Item in an order')]
final class OrderItem
{
    #[GraphQLField(type: 'Int', description: 'Item ID')]
    public int $id;

    #[GraphQLField(type: 'Product', description: 'Product ordered')]
    public Product $product;

    #[GraphQLField(type: 'Int', description: 'Quantity ordered')]
    public int $quantity;

    #[GraphQLField(type: 'Float', description: 'Price paid for this item')]
    public float $pricePaid;
}

#[GraphQLType(name: 'Customer', description: 'Customer information')]
final class Customer
{
    #[GraphQLField(type: 'Int', description: 'Customer ID')]
    public int $id;

    #[GraphQLField(type: 'String', description: 'Customer name')]
    public string $name;

    #[GraphQLField(type: 'String', description: 'Email address')]
    public string $email;

    #[GraphQLField(type: 'String', nullable: true, description: 'Phone number')]
    public ?string $phone;
}

#[GraphQLType(name: 'Order', description: 'Customer order')]
final class Order
{
    #[GraphQLField(type: 'Int', description: 'Order ID')]
    public int $id;

    #[GraphQLField(type: 'Customer', description: 'Customer who placed order')]
    public Customer $customer;

    #[GraphQLField(type: 'String', description: 'Order status')]
    public string $status;

    #[GraphQLField(type: 'Float', description: 'Total order amount')]
    public float $total;

    #[GraphQLField(type: 'String', description: 'Order date')]
    public string $createdAt;
}

// Example usage with builder
function demonstrateEcommerceSchema(): void
{
    echo "=== FraiseQL PHP E-Commerce Schema Example ===\n\n";

    // Register attribute-based types
    echo "Step 1: Registering attribute-based types...\n";
    StaticAPI::register(Category::class);
    StaticAPI::register(Product::class);
    StaticAPI::register(OrderItem::class);
    StaticAPI::register(Customer::class);
    StaticAPI::register(Order::class);
    echo "✓ Registered 5 attribute-based types\n\n";

    // Create Query and Mutation types using TypeBuilder
    echo "Step 2: Building Query type with fluent API...\n";
    $queryBuilder = \FraiseQL\TypeBuilder::type('Query')
        ->description('Root query type for e-commerce API')
        ->field('products', 'Product', isList: true, description: 'List all products')
        ->field('product', 'Product', nullable: true, description: 'Get product by ID')
        ->field('categories', 'Category', isList: true, description: 'List all categories')
        ->field('orders', 'Order', isList: true, description: 'List orders for customer')
        ->field('order', 'Order', nullable: true, description: 'Get order details');

    echo "✓ Query type has " . $queryBuilder->getFieldCount() . " fields\n\n";

    // Create Mutation type
    echo "Step 3: Building Mutation type...\n";
    $mutationBuilder = \FraiseQL\TypeBuilder::type('Mutation')
        ->description('Root mutation type for e-commerce API')
        ->field('createOrder', 'Order', description: 'Create a new order')
        ->field('updateProduct', 'Product', nullable: true, description: 'Update product')
        ->field('cancelOrder', 'Boolean', description: 'Cancel an order')
        ->field('addToCart', 'Boolean', description: 'Add item to shopping cart');

    echo "✓ Mutation type has " . $mutationBuilder->getFieldCount() . " fields\n\n";

    // Format and export schema
    echo "Step 4: Formatting complete schema...\n";
    $registry = \FraiseQL\SchemaRegistry::getInstance();
    $formatter = new SchemaFormatter();

    $schema = $formatter->formatBuilders(
        $queryBuilder,
        $mutationBuilder
    );

    echo "✓ Schema created with " . $schema->getTypeCount() . " root types\n\n";

    // Combine attribute types with builders
    echo "Step 5: Exporting complete schema...\n";
    $attributeSchema = $formatter->formatRegistry(
        $registry,
        description: 'Complete e-commerce GraphQL schema'
    );

    $json = $attributeSchema->toJson();
    echo "Schema exported successfully!\n";
    echo "Total schema size: " . strlen($json) . " bytes\n";
    echo "Total types: " . $attributeSchema->getTypeCount() . "\n";
    echo "Scalars: " . implode(', ', $attributeSchema->getScalarNames()) . "\n\n";

    // Inspect specific types
    echo "Step 6: Analyzing schema structure...\n";
    echo "Product type fields:\n";
    $productFields = StaticAPI::getTypeFields('Product');
    foreach ($productFields as $field) {
        $type = $field->getGraphQLTypeString();
        $description = $field->description ?? '(no description)';
        echo "  - {$field->name}: {$type}\n    {$description}\n";
    }
    echo "\n";

    // Show order of types
    echo "Step 7: Registered type names:\n";
    $typeNames = StaticAPI::getTypeNames();
    echo "Types: " . implode(', ', $typeNames) . "\n";
    echo "Total: " . count($typeNames) . " types\n\n";

    // Display sample JSON
    echo "Step 8: Sample JSON export (first 2 types):\n";
    $schemaArray = $attributeSchema->toArray();
    $sampleTypes = array_slice($schemaArray['types'], 0, 2, true);
    echo json_encode(
        [
            'version' => $schemaArray['version'],
            'description' => $schemaArray['description'] ?? null,
            'types' => $sampleTypes,
            'scalars' => $schemaArray['scalars'],
        ],
        JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES
    ) . "\n";
}

// Only run when executed directly
if (basename(__FILE__) === basename($_SERVER['SCRIPT_NAME'] ?? '')) {
    try {
        demonstrateEcommerceSchema();
    } catch (\Exception $e) {
        echo "Error: " . $e->getMessage() . "\n";
        exit(1);
    }
}

File: fraiseql-java/src/main/java/com/fraiseql/examples/EcommerceSchema.java
package com.fraiseql.examples;

import com.fraiseql.core.*;

/**
 * EcommerceSchema - Advanced example with product catalog, shopping cart, and orders
 *
 * This example demonstrates:
 * 1. Complex types with multiple fields
 * 2. Type relationships (Product, Cart, Order, etc.)
 * 3. Advanced queries with filtering arguments
 * 4. Mutations for shopping operations
 * 5. Real-world GraphQL schema design
 *
 * Output: ecommerce-schema.json - Ready for fraiseql-cli compile
 *
 * Usage:
 *   mvn exec:java -Dexec.mainClass="com.fraiseql.examples.EcommerceSchema"
 */
public class EcommerceSchema {

    public static void main(String[] args) {
        try {
            System.out.println("FraiseQL Java - Ecommerce Schema Example");
            System.out.println("========================================\n");

            // Register types
            System.out.println("1. Registering ecommerce types...");
            FraiseQL.registerTypes(
                Product.class,
                Category.class,
                Customer.class,
                CartItem.class,
                Order.class,
                OrderItem.class,
                Review.class
            );
            System.out.println("   ✓ 7 types registered\n");

            // Register queries
            System.out.println("2. Registering queries...");

            FraiseQL.query("products")
                .returnType(Product.class)
                .returnsArray(true)
                .arg("categoryId", "Int")
                .arg("limit", "Int")
                .arg("offset", "Int")
                .description("Get products with optional category and pagination")
                .register();
            System.out.println("   ✓ products");

            FraiseQL.query("product")
                .returnType(Product.class)
                .arg("id", "Int")
                .description("Get a specific product")
                .register();
            System.out.println("   ✓ product");

            FraiseQL.query("categories")
                .returnType(Category.class)
                .returnsArray(true)
                .description("Get all product categories")
                .register();
            System.out.println("   ✓ categories");

            FraiseQL.query("customer")
                .returnType(Customer.class)
                .arg("id", "Int")
                .description("Get customer profile")
                .register();
            System.out.println("   ✓ customer");

            FraiseQL.query("orders")
                .returnType(Order.class)
                .returnsArray(true)
                .arg("customerId", "Int")
                .arg("limit", "Int")
                .description("Get customer's orders")
                .register();
            System.out.println("   ✓ orders");

            FraiseQL.query("reviews")
                .returnType(Review.class)
                .returnsArray(true)
                .arg("productId", "Int")
                .description("Get reviews for a product")
                .register();
            System.out.println("   ✓ reviews\n");

            // Register mutations
            System.out.println("3. Registering mutations...");

            FraiseQL.mutation("createCustomer")
                .returnType(Customer.class)
                .arg("name", "String")
                .arg("email", "String")
                .arg("phone", "String")
                .description("Create a new customer account")
                .register();
            System.out.println("   ✓ createCustomer");

            FraiseQL.mutation("updateCustomer")
                .returnType(Customer.class)
                .arg("id", "Int")
                .arg("name", "String")
                .arg("email", "String")
                .description("Update customer profile")
                .register();
            System.out.println("   ✓ updateCustomer");

            FraiseQL.mutation("addToCart")
                .returnType(CartItem.class)
                .arg("customerId", "Int")
                .arg("productId", "Int")
                .arg("quantity", "Int")
                .description("Add item to shopping cart")
                .register();
            System.out.println("   ✓ addToCart");

            FraiseQL.mutation("removeFromCart")
                .returnType(CartItem.class)
                .arg("customerId", "Int")
                .arg("productId", "Int")
                .description("Remove item from shopping cart")
                .register();
            System.out.println("   ✓ removeFromCart");

            FraiseQL.mutation("checkout")
                .returnType(Order.class)
                .arg("customerId", "Int")
                .arg("shippingAddress", "String")
                .description("Checkout and create order")
                .register();
            System.out.println("   ✓ checkout");

            FraiseQL.mutation("createReview")
                .returnType(Review.class)
                .arg("productId", "Int")
                .arg("customerId", "Int")
                .arg("rating", "Int")
                .arg("text", "String")
                .description("Create a product review")
                .register();
            System.out.println("   ✓ createReview\n");

            // Export schema
            System.out.println("4. Exporting schema to ecommerce-schema.json...");
            FraiseQL.exportSchema("ecommerce-schema.json");
            System.out.println("   ✓ Schema exported\n");

            // Print summary
            SchemaRegistry registry = FraiseQL.getRegistry();
            System.out.println("Ecommerce Schema Summary:");
            System.out.println("------------------------");
            System.out.println("Types:     " + registry.getAllTypes().size());
            System.out.println("Queries:   " + registry.getAllQueries().size());
            System.out.println("Mutations: " + registry.getAllMutations().size());
            System.out.println("\nTypes:");
            for (String typeName : registry.getAllTypes().keySet()) {
                System.out.println("  - " + typeName);
            }

        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
            System.exit(1);
        }
    }

    @GraphQLType(description = "A product in the catalog")
    public static class Product {
        @GraphQLField(description = "Product ID")
        public int id;

        @GraphQLField(description = "Product name")
        public String name;

        @GraphQLField(description = "Product description")
        public String description;

        @GraphQLField(description = "Price in cents")
        public int price;

        @GraphQLField(description = "Category ID")
        public int categoryId;

        @GraphQLField(description = "Stock quantity")
        public int stock;

        @GraphQLField(name = "created_at", description = "When product was added")
        public String createdAt;
    }

    @GraphQLType(description = "A product category")
    public static class Category {
        @GraphQLField(description = "Category ID")
        public int id;

        @GraphQLField(description = "Category name")
        public String name;

        @GraphQLField(description = "Number of products in category")
        public int productCount;
    }

    @GraphQLType(description = "A customer account")
    public static class Customer {
        @GraphQLField(description = "Customer ID")
        public int id;

        @GraphQLField(description = "Customer name")
        public String name;

        @GraphQLField(description = "Customer email")
        public String email;

        @GraphQLField(description = "Customer phone")
        public String phone;

        @GraphQLField(description = "Total spent in cents")
        public int totalSpent;

        @GraphQLField(name = "created_at", description = "Account creation date")
        public String createdAt;
    }

    @GraphQLType(description = "Item in a shopping cart")
    public static class CartItem {
        @GraphQLField(description = "Cart item ID")
        public int id;

        @GraphQLField(description = "Customer ID")
        public int customerId;

        @GraphQLField(description = "Product ID")
        public int productId;

        @GraphQLField(description = "Quantity in cart")
        public int quantity;

        @GraphQLField(description = "Added to cart at")
        public String addedAt;
    }

    @GraphQLType(description = "A customer order")
    public static class Order {
        @GraphQLField(description = "Order ID")
        public int id;

        @GraphQLField(description = "Customer ID")
        public int customerId;

        @GraphQLField(description = "Total price in cents")
        public int total;

        @GraphQLField(description = "Order status (pending, processing, shipped, delivered)")
        public String status;

        @GraphQLField(description = "Shipping address")
        public String shippingAddress;

        @GraphQLField(name = "created_at", description = "Order date")
        public String createdAt;

        @GraphQLField(name = "shipped_at", description = "Ship date")
        public String shippedAt;
    }

    @GraphQLType(description = "Item in an order")
    public static class OrderItem {
        @GraphQLField(description = "Order item ID")
        public int id;

        @GraphQLField(description = "Order ID")
        public int orderId;

        @GraphQLField(description = "Product ID")
        public int productId;

        @GraphQLField(description = "Quantity ordered")
        public int quantity;

        @GraphQLField(description = "Price at time of order")
        public int price;
    }

    @GraphQLType(description = "Product review")
    public static class Review {
        @GraphQLField(description = "Review ID")
        public int id;

        @GraphQLField(description = "Product ID")
        public int productId;

        @GraphQLField(description = "Customer ID")
        public int customerId;

        @GraphQLField(description = "Star rating (1-5)")
        public int rating;

        @GraphQLField(description = "Review text")
        public String text;

        @GraphQLField(name = "created_at", description = "Review date")
        public String createdAt;

        @GraphQLField(description = "Number of helpful votes")
        public int helpfulVotes;
    }
}
